File: verify/timing.py
"""Closed-form cost per evaluation, single thread, 10 000 RTs at one parameter set.
Same parameter set as verify/timing.R.  Needs the compiled kernel (cythonize -3 -i src/ddm_kernel.pyx)."""
import sys; sys.path.insert(0, "src")
import numpy as np, timeit
import ddm_kernel as K, ddm_fast as F
t = np.random.default_rng(1).uniform(0.1, 3, 10000)
a, v, w1, w2, sv, st0 = 1.2, 1.0, 0.4, 0.6, 1.0, 0.2
us = lambda fn: 1e6 * min(timeit.repeat(fn, number=1, repeat=5)) / len(t)
print(f"closed form, compiled  {'six-parameter density':38s} {us(lambda: K.g_full_sz(t, v, sv, a, w1, w2)):8.2f} us/eval")
print(f"closed form, numpy     {'six-parameter density + five gradients':38s} {us(lambda: F.grad_full_sz(t, v, sv, a, w1, w2)):8.2f} us/eval")
print(f"closed form, compiled  {'seven-parameter density':38s} {us(lambda: K.f7(t, v, sv, a, w1, w2, 0.0, st0)):8.2f} us/eval")

File: verify/timing.R
# WienR cost per evaluation, single thread, 10 000 RTs at one parameter set (the fitting case).
# Same parameter set as verify/timing.py.  Run from anywhere: Rscript verify/timing.R
suppressMessages(library(WienR)); set.seed(1); n <- 10000
t <- runif(n, 0.1, 3); a <- 1.2; v <- 1.0; w <- 0.5; sv <- 1.0; sw <- 0.2; st0 <- 0.2
us <- function(f, reps = 5) { best <- Inf; for (i in 1:reps) { t0 <- Sys.time(); f(); best <- min(best, as.numeric(Sys.time() - t0, units = "secs")) }; 1e6 * best / n }
cat(sprintf("WienR %-42s %8.2f us/eval\n", "sv only (its closed form, Blurton 2017)", us(function() WienerPDF(t, "lower", a, v, w, sv = sv, sw = 0, precision = 1e-12))))
cat(sprintf("WienR %-42s %8.2f us/eval\n", "six-parameter density (sv, sw)", us(function() WienerPDF(t, "lower", a, v, w, sv = sv, sw = sw, precision = 1e-12, n.evals = 0))))
cat(sprintf("WienR %-42s %8.2f us/eval\n", "five gradients (sv, sw)", us(function() for (f in c(dvWienerPDF, dsvWienerPDF, daWienerPDF, dwWienerPDF, dswWienerPDF)) f(t, "lower", a, v, w, sv = sv, sw = sw, precision = 1e-12, n.evals = 0))))
cat(sprintf("WienR %-42s %8.2f us/eval\n", "seven-parameter density (sv, sw, st0)", us(function() WienerPDF(t, "lower", a, v, w, sv = sv, sw = sw, st0 = st0, precision = 1e-12, n.evals = 0), reps = 2)))
suppressMessages(library(rtdists))   # default precision = 3; absolute z = a*w, sz = a*sw
cat(sprintf("rtdists %-40s %8.2f us/eval\n", "six-parameter density (precision 3)", us(function() ddiffusion(t, "lower", a = a, v = v, t0 = 0, z = w * a, sz = sw * a, sv = sv, st0 = 0, s = 1, precision = 3))))
cat(sprintf("rtdists %-40s %8.2f us/eval\n", "seven-parameter density (precision 3)", us(function() ddiffusion(t, "lower", a = a, v = v, t0 = 0, z = w * a, sz = sw * a, sv = sv, st0 = st0, s = 1, precision = 3), reps = 2)))
